runtime/prng.c

```diff
@@ -0,0 +1,102 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy, projet Cambium, College de France and Inria     */
+/*                                                                        */
+/*   Copyright 2021 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include "caml/alloc.h"
+#include "caml/bigarray.h"
+#include "caml/mlvalues.h"
+
+/* The L64X128 member of the LXM family.  Taken from figure 1 in
+   "LXM: Better Splittable Pseudorandom Number Generators
+    (and Almost as Fast)" by Guy L. Steele Jr. and Sebastiano Vigna,
+    OOPSLA 2021.  */
+
+static const uint64_t M = 0xd1342543de82ef95;
+
+struct LXM_state {
+  uint64_t a;            /* per-instance additive parameter (odd) */
+  uint64_t s;            /* state of the LCG subgenerator */
+  uint64_t x[2];         /* state of the XBG subgenerator (not 0) */
+};
+
+/* In OCaml, states are represented as a 1D big array of 64-bit integers */
+
+#define LXM_val(v) ((struct LXM_state *) Caml_ba_data_val(v))
+
+static inline uint64_t rotl(const uint64_t x, int k) {
+  return (x << k) | (x >> (64 - k));
+}
+
+CAMLprim uint64_t caml_lxm_next_unboxed(value v)
+{
+  uint64_t z, q0, q1;
+  struct LXM_state * st = LXM_val(v);
+
+  /* Combining operation */
+  z = st->s + st->x[0];
+  /* Mixing function */
+  z = (z ^ (z >> 32)) * 0xdaba0b6eb09322e3;
+  z = (z ^ (z >> 32)) * 0xdaba0b6eb09322e3;
+  z = (z ^ (z >> 32));
+  /* LCG update */
+  st->s = st->s * M + st->a;
+  /* XBG update */
+  q0 = st->x[0]; q1 = st->x[1];
+  q1 ^= q0;
+  q0 = rotl(q0, 24);
+  q0 = q0 ^ q1 ^ (q1 << 16);
+  q1 = rotl(q1, 37);
+  st->x[0] = q0; st->x[1] = q1;
+  /* Return result */
+  return z;
+}
+
+CAMLprim value caml_lxm_next(value v)
+{
+  return caml_copy_int64(caml_lxm_next_unboxed(v));
+}
+
+static void caml_lxm_set(value v, uint64_t i1, uint64_t i2,
+                                  uint64_t i3, uint64_t i4)
+{
+  struct LXM_state * st = LXM_val(v);
+  st->a = i1 | 1;              /* must be odd */
+  st->x[0] = i2 != 0 ? i2 : 1; /* must be nonzero */
+  st->x[1] = i3 != 0 ? i3 : 2; /* must be nonzero */
```

Comment:
The paper writes something different about the `x` parameter initialization.

> The constructor then forces the additive parameter to be odd by setting its low-order
bit to 1, but beyond that no additional vetting of the additive parameter (to reject “weak valuesž
[Steele, Lea, and Flood 2014]) is necessary. In the unlikely circumstance that the state for the XBG
subgenerator is entirely 0, it is necessary to force it to be nonzero; this could be done by making
additional calls to nextLong() or nextInt(), but it is easier (and acceptable in practice, because it
happens so rarely) to derive the new XBG state from the new LCG state.

I'm not sure what they mean by "derive the new XBG state from the new LCG state", but maybe they are thinking of setting one of x[0] or x[1] to the odd `st->a` value? (Note: the code currently suggests that neither of x0 and x1 can be zero, but in fact the paper says that having only one of them non-zero is enough.)

Reply:
Instead of "x[0] and x[1] must not be both 0", I used the more conservative "none of x[0] and x[1] should be zero" because that's what the Xoshiro papers recommend, if I remember correctly.

For states that are created by splitting, `i2` and `i3` were obtained by pseudo-random number generation, so the probability that they are 0 is extremely low, and using 1 and 2 in this unlikely case is unnoticeable.

What they mean by "derive the new XBG state from the new LCG state" is better explained in the Xoshiro papers.  To initialize an XBG from a single 64-bit integer seed, Vigna suggests to use... a SplitMix PRNG, initialized with this seed, and run N times to get N good-looking numbers to initialize the N-word state of the XBG.

For states that are created by initialization from an array of integers (function `Random.full_init` and related functions), the situation is more delicate.  Sometimes we are given a single integer (function `Random.init`), in which case Vigna's suggestion would apply.  But sometimes we are given 12 bytes with good entropy (function `Random.self_init` when `/dev/urandom` can be read) and we should "spread out" this 96-bit entropy on the 4 components of the L64X128 state.

After some thinking, I implemented a cute solution where the integer array is hashed (as a string of 64-bit "characters") using a hash function that produces a 256-bit result, which is then taken as the initial L64X128 state.  It is hoped that the result incorporates all the entropy present in the integer array, up to 256 bits, yet produces four 64-bit numbers that don't have strong correlations and "look random already" (whatever that means).

This solution is commit 83c283086 in this PR.

Reply:
I understand the solution and I think it should work fine indeed. Two remarks (that don't suggest any change):
- We would also have the choice of asking /dev/urandom for more input (but that doesn't work when /dev/urandom is not available, where we generate a fixed amount of bytes with bad entropy)
- We already have a hash function implemented in the runtime, Murmurhash 3 (the 32bit-producing version). I understand that you need a 256-large output, so FNV1a fits the bill better.

Reply:
Hash functions come in all sizes and shapes :-)  However, I woke up in the middle of the night with the realization that FNV1a doesn't avalanche much on short inputs (e.g. a single integer).

So, I just pushed yet another proposal that uses good old MD5 as the hashing function.  It only produces 128 bits of output, so I call it twice with some diversification of the input.  Now, it should avalanche perfectly, while preserving up to 128 bits of entropy present in the seed.  It's a bit slower than the previous implementation, but still significantly faster than the one in OCaml today.

As an additional benefit, all the initialization code is now done in OCaml, reducing the C code in runtime/prng.c to the one speed-critical function `next`.

Reply:
Further remark 1: yes, we need a 256-bit (or more) crypto hash function to replace MD5 in Digest.  It's on my to do list.

Further remark 2: yes, we should revisit the PRNG initialization from /dev/urandom.  This will come next.

Reply:
Regarding `next`, I found some time again to try a pure OCaml implementation (using bigarrays this time), and in optimal conditions (amd64 backend, native mode) it outperforms the C version by ~5% in microbenchmarks.
It is less readable than the C version though, and of course it will probably be significantly slower in bytecode and 32-bit native backends.

Reply:
I'm not surprised you're getting good performance from a pure OCaml implementation on x86-64: there is not much to optimize in this code, and the only optimization OCaml misses is the recognition of the "rotate" instruction.  Note however that I forgot to declare the C primitive as "noalloc".  This would regain a few percents...